README.md

```diff
@@ -59,20 +59,34 @@ If you want to automatically search root modules in your workspace and exclude s
 ### Formatting
 To enable formatting, it is recommended that the following be added to the extension settings for the Terraform extension:
 ```
-    "[terraform]": {
-        "editor.formatOnSave": true
-    }
+"[terraform]": {
+  "editor.defaultFormatter": "hashicorp.terraform",
+  "editor.formatOnSave": true,
+  "editor.formatOnSaveMode": "file"
+}
 ```
 
-It is also recommended to set a default formatter if you have other extensions installed which also claim a capability to format Terraform files:
+It is recommended to set `editor.defaultFormatter` to ensure that VS Code knows which extension to use to format your files. It is possible to have more than one extension installed which claim a capability to format Terraform files.
+
+When using the `editor.formatOnSaveMode` setting, only `file` is currently supported. The `modifications` or `modificationsIfAvailable` settings [use the currently configured SCM](https://code.visualstudio.com/updates/v1_49#_only-format-modified-text) to detect file line ranges that have changed and send those ranges to the formatter. The `file` setting works because `terraform fmt` was written to format an entire file, not ranges. If you don't have a SCM enabled for the files you are editing, `modifications` won't work at all. The `modificationsIfAvailable` setting will fall back to `file` if there is no SCM and will appear to work sometimes.
```

Comment:
> The `file` setting works because `terraform fmt` was written to format an entire file, not ranges.

You're not wrong here in the sense that it was _designed for_ whole files but there's no strong reason why we can't implement [`textDocument/rangeFormatting`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_rangeFormatting) and pass just part of the config to stdin to `terraform fmt -`.

I guess this will expose some edge cases, like formatting range which slices through an attribute or block - and I'm unsure how the underlying implementation deals with that or how it should.

Anyway - I'm happy with your conclusion here, just wanted to say that we can address this on LS side if there's interest.

Reply:
Thanks for the added context. The purpose of this text is to instruct the user how to use it as currently implemented so I did not include the possibility of passing ranges of text to terraform to format. Since we both thought of this, this sounds like I should create a ticket for this idea.